code/modules/mob/living/carbon/human/human.dm

```diff
@@ -650,46 +650,68 @@
 				to_chat(src, "<span class='warning'>\The [S] pulls \the [hand] from your grip!</span>")
 	rad_act(current_size * 3)
 
-/mob/living/carbon/human/proc/do_cpr(mob/living/carbon/C)
-	CHECK_DNA_AND_SPECIES(C)
+#define CPR_PANIC_SPEED (0.8 SECONDS)
 
-	if(C.stat == DEAD || (HAS_TRAIT(C, TRAIT_FAKEDEATH)))
-		to_chat(src, "<span class='warning'>[C.name] is dead!</span>")
-		return
-	if(is_mouth_covered())
-		to_chat(src, "<span class='warning'>Remove your mask first!</span>")
-		return 0
-	if(C.is_mouth_covered())
-		to_chat(src, "<span class='warning'>Remove [p_their()] mask first!</span>")
-		return 0
-
-	if(C.cpr_time < world.time + 30)
-		visible_message("<span class='notice'>[src] is trying to perform CPR on [C.name]!</span>", \
-						"<span class='notice'>You try to perform CPR on [C.name]... Hold still!</span>")
-		if(!do_mob(src, C))
-			to_chat(src, "<span class='warning'>You fail to perform CPR on [C]!</span>")
-			return 0
-
-		var/they_breathe = !HAS_TRAIT(C, TRAIT_NOBREATH)
-		var/they_lung = C.getorganslot(ORGAN_SLOT_LUNGS)
-
-		if(C.health > C.crit_threshold)
-			return
+/// Performs CPR on the target after a delay.
+/mob/living/carbon/human/proc/do_cpr(mob/living/carbon/target)
+	var/panicking = FALSE
+
+	while (TRUE)
```

Comment:
Instead of `while(TRUE)` use a `do-while(actual condition that would end the proc)`

Reply:
I know what do-while is, but there's several conditions at which the CPR should stop. I'm not sure what's wrong with this.

Reply:
We have a `goto` instruction as well, but same as `while(TRUE)` which is not expected to run forever, it's generally considered bad practice.

The behavior you are looking for here is to run the code once at least, and break on specific conditions. And if others are met, to run again. That is exactly what `do-while` does. `while(TRUE)` offers much less clarity.

The fact that something works doesn't mean is good, and you can perfectly write good code logic on a `do-while` with breaks inside of it. You just need to define a condition at the end of it that can properly assess whether the code should be ran again or not, not obscured inside the loop. It seems to me that the `panicking` condition is a good candidate for it.

This is about structuring sound code logic, not just making it work.

Reply:
> It seems to me that the panicking condition is a good candidate for it.

It's not? The panicking variable is there to tell the CPR that it needs to go faster, not that it should stop.

There are several reasons a CPR should fail that would look dumb, and impractical, to combine into one long conditional.

Reply:
Choose whatever condition you want to be evaluated at the end of the loop, to decide whether to continue or not, if not `panicking`. It doesn't need to be one long conditional, it just needs to be one different from `TRUE`.

Structure your code in a clear and readable way and avoid `while(TRUE)` unless you are actually running an infinite loop.

Reply:
I'll see what I can do, yeah.